crates/sui-indexer-alt-consistent-store/src/store/synchronizer.rs

```diff
@@ -99,28 +98,35 @@ impl Synchronizer {
         let mut queue = Queue::new();
         let mut tasks = Vec::new();
 
-        let pre_snap = Arc::new(Barrier::new(self.last_checkpoints.len()));
-        let post_snap = Arc::new(Barrier::new(self.last_checkpoints.len()));
+        let pre_snap = Arc::new(Barrier::new(self.last_watermarks.len()));
+        let post_snap = Arc::new(Barrier::new(self.last_watermarks.len()));
 
-        // All tasks will arrange to take a snapshot before writing the next checkpoint that is new
-        // to all pipelines (allowing all pipelines to catch up with each other).
-        let mut next_snapshot_checkpoint = 0;
-        for (pipeline, last_checkpoint) in self.last_checkpoints {
-            let (tx, rx) = mpsc::channel(self.buffer_size);
-            let next_checkpoint = last_checkpoint
-                .map(|v| v + 1)
-                .unwrap_or(self.first_checkpoint);
+        // Calculate the first checkpoint we will have data for, across all pipelines.
+        let Some(init_checkpoint) = self
+            .last_watermarks
+            .values()
+            .map(|w| w.map_or(self.first_checkpoint, |w| w.checkpoint_hi_inclusive))
+            .max()
+        else {
+            bail!("No pipelines registered with the synchronizer");
+        };
+
+        // The next snapshot should be taken at the next stride boundary after that initial
+        // checkpoint.
+        let next_snapshot_checkpoint = ((init_checkpoint / self.stride) + 1) * self.stride;
```

Comment:
This change is what fixes the bug -- instead of calculating the `next_snapshot_checkpoint` in the same loop we create tasks in, we perform one loop to figure out where the next snapshot checkpoint should be, and then another to start the tasks.